orchestra/executor.py

```diff
@@ -256,6 +262,100 @@ def _remove_satisfied_attracting_components(graph):
+        so A@2 would not find the exact build it was expecting.
+
+        The pass transforms the graph above into:
+
+               +-------+
+            +--+  A@1  +--+
+            |  +-------+  |
+        +---v---+     +---v---+
+        |  B@1  +----->  A@2  +--+
+        +---+---+     +-------+  |
+            |                +---v---+
+            +---------------->  B@2  |
+                             +-------+
+
+        For each component C the algorithm creates a list of groups of actions, one for each build.
+        Each group contains:
+         1. actions that pertain to a specific build of the component
+         2. actions that directly depend on actions of point 1
+        The algorithm tries all possible permutations of the groups in the list.
+        For each permutation [G1, G2, ..., Gn] all actions in
+        group Gi are marked to depend on all actions in group Gi+1.
+        The graph is checked for cycles and if none are found the order is accepted.
+        """
+        scheduled_actions_per_build = defaultdict(set)
+        scheduled_builds_per_component = defaultdict(set)
+        scheduled_actions_per_direct_build_dependency = defaultdict(set)
+
+        for action in dependency_graph.nodes:
+            if isinstance(action, ActionForBuild):
+                scheduled_builds_per_component[action.component].add(action.build)
+                scheduled_actions_per_build[action.build].add(action)
+                for d in dependency_graph.predecessors(action):
+                    scheduled_actions_per_direct_build_dependency[action.build].add(d)
+
+        groups_by_component = defaultdict(list)
+        for c, blds in scheduled_builds_per_component.items():
+            if len(blds) < 2:
+                continue
+
+            for bld in blds:
+                group = set()
+                for act in scheduled_actions_per_build[bld]:
+                    group.add(act)
+                for act in scheduled_actions_per_direct_build_dependency[bld]:
+                    group.add(act)
+                groups_by_component[c].append(group)
+
+        for component, group in groups_by_component.items():
+            dependency_graph = self._try_group_orders(dependency_graph, group)
+            if dependency_graph is None:
+                raise OrchestraException(f"Could not enforce an order between actions "
+                                         f"of component {component} pertaining to multiple builds")
+
+        return dependency_graph
+
+    @staticmethod
+    def _try_group_orders(dependency_graph, group):
+        for permutation in permutations(group):
+            # TODO: duplicating the graph is not good for performance,
+            #  might be worth removing nodes manually
+            depgraph_copy = dependency_graph.copy()
+
+            for i in range(len(permutation) - 1):
+                cur_group = permutation[i]
+                next_group = permutation[i + 1]
```

Comment:
You can likely use `zip` and be more pythonic here, but nevermind.

Reply:
If you are ok with
```
for g1, g2 in zip(permutation, permutation[1:]):
    ...
```
then I'm all for it, I thought it would have been a bit obscure